File: 150-evaluate-reverse-polish-notation/evaluate-reverse-polish-notation.ts
function evalRPN(tokens: string[]): number {
    const operations = new Map<string, (a: number, b: number) => number>([
        ['+', (a: number, b: number) => a + b],
        ['-', (a: number, b: number) => a - b],
        ['*', (a: number, b: number) => a * b],
        ['/', (a: number, b: number) => Math.trunc(a / b)],
    ])
    const stack = new Array<string>();
    for(let char of tokens){
        if(operations.has(char)){
            const a = parseInt(stack.pop());
            const b = parseInt(stack.pop());
            const ans = operations.get(char)!(b,a);
            stack.push(ans.toString());
        }
        else stack.push(char);
    }
    return parseInt(stack.pop());
};

File: 150-evaluate-reverse-polish-notation/evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Deque<String> stack = new ArrayDeque<>();
        for(String curToken : tokens) {
            if(
                curToken.equals("+") ||
                curToken.equals("-") ||
                curToken.equals("*") ||
                curToken.equals("/") 
                ){
                    int secondOperand = Integer.parseInt(stack.poll());
                    int firstOperand = Integer.parseInt(stack.poll());
                    int ans = evaluate(firstOperand , secondOperand , curToken);
                    stack.push(ans + "");
                }
                else stack.push(curToken);
        }
        return Integer.parseInt(stack.poll());
    }

    private int evaluate(int a , int b , String operator){
        switch(operator){
            case "+": return a + b;
            case "-": return a - b;
            case "*": return a * b;
            case "/": return a / b;
            default: throw new IllegalArgumentException();
        }
    }
}